

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Visium Kidney &mdash; Giotto 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/togglebutton.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/sphinxcontrib-images/LightBox2/lightbox2/css/lightbox.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
        <script src="../../_static/togglebutton.js"></script>
        <script >var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
        <script src="../../_static/sphinxcontrib-images/LightBox2/lightbox2/js/jquery-1.11.0.min.js"></script>
        <script src="../../_static/sphinxcontrib-images/LightBox2/lightbox2/js/lightbox.min.js"></script>
        <script src="../../_static/sphinxcontrib-images/LightBox2/lightbox2-customize/jquery-noconflict.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html">
          

          
            
            <img src="../../_static/GiottoLogo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">General Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">How to Get Started With Giotto</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../giottoworkflowanalyses.html">Giotto Workflow</a></li>
</ul>
<p class="caption"><span class="caption-text">Giotto Example Datasets</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../datasets.html">Mini Datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../datasets.html#full-datasets">Full Datasets</a></li>
</ul>
<p class="caption"><span class="caption-text">How Tos and Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/giottoclass.html">Working with Giotto Class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/howtosubsetgiottoobject.html">How to Subset Giotto</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/howtovisualizeandsaveplots.html">How to Visualize Giotto Object</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/testandstoremultipleanayses.html">How to Test and Store Multiple</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/waystovisualizespatialdata.html">Ways to Visualize</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/workingwithimages.html">Working with Images</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tipsandtricks/visualizewithvoronoiplots.html">Visulize Voronoi</a></li>
</ul>
<p class="caption"><span class="caption-text">Supplemental Information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../errorsfaqsandtips.html">Errors and FAQS</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../additionalinformation.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../additionalinformation.html#how-to-contribute">How to contribute?</a></li>
<li class="toctree-l1"><a class="reference external" href="https://​github.com/​RubD/​Giotto/​">Browse Source Code</a></li>
<li class="toctree-l1"><a class="reference external" href="https://​github.com/​RubD/​Giotto/​issues">Report a Bug</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Giotto</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Visium Kidney</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/subsections/datasets/visium_kidney.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="visium-kidney">
<h1>Visium Kidney<a class="headerlink" href="#visium-kidney" title="Permalink to this headline">¶</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This tutorial was written with <strong>Giotto version 0.3.6.9046</strong>, your version is <strong>1.0.3</strong>. This is a more recent version and results should be reproducible.</p>
</div>
<div class="section" id="install-python-and-r-modules">
<h2>Install Python and R Modules<a class="headerlink" href="#install-python-and-r-modules" title="Permalink to this headline">¶</a></h2>
<p>To run this vignette you need to install all the necessary Python modules. This can be done either (1) manually or (2) via our Installation Tool (from within R).</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="../../gettingstarted.html#manualinstallation"><span class="std std-ref">Click here for manual installation instructions</span></a></p></li>
<li><p><a class="reference internal" href="../../gettingstarted.html#automaticinstallation"><span class="std std-ref">Click here for our automatic installation tool instructions</span></a></p></li>
</ol>
</div>
<div class="section" id="optional-set-giotto-instructions">
<h2><em>Optional: Set Giotto Instructions</em><a class="headerlink" href="#optional-set-giotto-instructions" title="Permalink to this headline">¶</a></h2>
<p><em>Within R</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># to automatically save figures in save_dir set save_plot to TRUE</span>
<span class="n">temp_dir</span> <span class="o">=</span> <span class="n">getwd</span><span class="p">()</span>
<span class="n">temp_dir</span> <span class="o">=</span> <span class="s1">&#39;~/Temp/&#39;</span>
<span class="n">myinstructions</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">=</span> <span class="n">temp_dir</span><span class="p">,</span>
                                      <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                                      <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="set-up-giotto">
<h2>Set-Up Giotto<a class="headerlink" href="#set-up-giotto" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Giotto</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="set-a-working-directory">
<h3>Set A Working Directory<a class="headerlink" href="#set-a-working-directory" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#results_folder = &#39;/path/to/directory/&#39;</span>
<span class="n">results_folder</span> <span class="o">=</span> <span class="s1">&#39;/Volumes/Ruben_Seagate/Dropbox (Personal)/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Results/Visium/Brain/201226_results//&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="set-a-giotto-python-path">
<h3>Set A Giotto Python Path<a class="headerlink" href="#set-a-giotto-python-path" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># set python path to your preferred python version path</span>
<span class="c1"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span>
<span class="n">python_path</span> <span class="o">=</span> <span class="n">NULL</span>
<span class="k">if</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">python_path</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">installGiottoEnvironment</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="dataset-explanation">
<h2>Dataset Explanation<a class="headerlink" href="#dataset-explanation" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://www.10xgenomics.com/spatial-transcriptomics">10X genomics</a> recently launched a new platform to obtain spatial expression data using a Visium Spatial Gene Expression slide. The Visium brain data to run this tutorial can be found <a class="reference external" href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Adult_Mouse_Brain">here</a>.</p>
<div class="dropdown admonition">
<p class="admonition-title">Visium Technology</p>
<a class="reference internal image-reference" href="../../_images/visium_technology.png"><img alt="Visium Technology" src="../../_images/visium_technology.png" style="width: 500px;" /></a>
</div>
<p>High resolution png from original tissue:</p>
<a class="reference internal image-reference" href="../../_images/mouse_kidney_highres.png"><img alt="mouse_kidney_highres.png" src="../../_images/mouse_kidney_highres.png" style="width: 400px;" /></a>
</div>
<div class="section" id="giotto-global-instructions-and-preparations">
<h2>1. Giotto Global Instructions and Preparations<a class="headerlink" href="#giotto-global-instructions-and-preparations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="create-instructions">
<h3>1.1. Create Instructions<a class="headerlink" href="#create-instructions" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">instrs</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span>
                          <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                          <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                          <span class="n">python_path</span> <span class="o">=</span> <span class="n">python_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="provide-path-to-visium-folder">
<h3>1.2. Provide Path to Visium Folder<a class="headerlink" href="#provide-path-to-visium-folder" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="s1">&#39;/path/to/Kidney_data/&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="create-giotto-object-and-process-the-data">
<h2>2. Create Giotto Object and Process The Data<a class="headerlink" href="#create-giotto-object-and-process-the-data" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## directly from visium folder</span>
<span class="n">visium_kidney</span> <span class="o">=</span> <span class="n">createGiottoVisiumObject</span><span class="p">(</span><span class="n">visium_dir</span> <span class="o">=</span> <span class="n">data_path</span><span class="p">,</span> <span class="n">expr_data</span> <span class="o">=</span> <span class="s1">&#39;raw&#39;</span><span class="p">,</span>
                                 <span class="n">png_name</span> <span class="o">=</span> <span class="s1">&#39;tissue_lowres_image.png&#39;</span><span class="p">,</span>
                                 <span class="n">gene_column_index</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">instructions</span> <span class="o">=</span> <span class="n">instrs</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_a_spatplot_image1.png"><img alt="2_a_spatplot_image.png" src="../../_images/2_a_spatplot_image1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check name</span>
<span class="n">showGiottoImageNames</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">)</span> <span class="c1"># &quot;image&quot; is the default name</span>
<span class="c1"># adjust parameters to align image (iterative approach)</span>
<span class="n">visium_kidney</span> <span class="o">=</span> <span class="n">updateGiottoImage</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">image_name</span> <span class="o">=</span> <span class="s1">&#39;image&#39;</span><span class="p">,</span>
                          <span class="n">xmax_adj</span> <span class="o">=</span> <span class="mi">1300</span><span class="p">,</span> <span class="n">xmin_adj</span> <span class="o">=</span> <span class="mi">1200</span><span class="p">,</span>
                          <span class="n">ymax_adj</span> <span class="o">=</span> <span class="mi">1100</span><span class="p">,</span> <span class="n">ymin_adj</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>

<span class="c1"># now it&#39;s aligned</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span> <span class="n">show_image</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_b_spatplot_image_adjusted&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_b_spatplot_image_adjusted1.png"><img alt="2_b_spatplot_image_adjusted.png" src="../../_images/2_b_spatplot_image_adjusted1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## check metadata</span>
<span class="n">pDataDT</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">)</span>

<span class="c1">## compare in tissue with provided jpg</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
        <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;0&#39;</span> <span class="o">=</span> <span class="s1">&#39;lightgrey&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span><span class="p">),</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_c_in_tissue&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_c_in_tissue1.png"><img alt="2_c_in_tissue.png" src="../../_images/2_c_in_tissue1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## subset on spots that were covered by tissue
metadata = pDataDT(visium_kidney)
in_tissue_barcodes = metadata[in_tissue == 1]$cell_ID
visium_kidney = subsetGiotto(visium_kidney, cell_ids = in_tissue_barcodes)

## filter
visium_kidney &lt;- filterGiotto(gobject = visium_kidney,
                      expression_threshold = 1,
                      gene_det_in_min_cells = 50,
                      min_det_genes_per_cell = 1000,
                      expression_values = c(&#39;raw&#39;),
                      verbose = T)

## normalize
visium_kidney &lt;- normalizeGiotto(gobject = visium_kidney, scalefactor = 6000, verbose = T)

## add gene &amp; cell statistics
visium_kidney &lt;- addStatistics(gobject = visium_kidney)
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_d_spatial_locations1.png"><img alt="2_d_spatial_locations.png" src="../../_images/2_d_spatial_locations1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">show_image</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
   <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;nr_genes&#39;</span><span class="p">,</span> <span class="n">color_as_factor</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
   <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_e_nr_genes&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_e_nr_genes1.png"><img alt="2_e_nr_genes.png" src="../../_images/2_e_nr_genes1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="dimension-reduction">
<h2>3. Dimension Reduction<a class="headerlink" href="#dimension-reduction" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## highly variable genes (HVG)</span>
<span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">calculateHVG</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
                      <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_a_HVGplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_a_HVGplot2.png"><img alt="3_a_HVGplot.png" src="../../_images/3_a_HVGplot2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## run PCA on expression values (default)</span>
<span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">runPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">scale_unit</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="n">screePlot</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">ncp</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_b_screeplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_b_screeplot2.png"><img alt="3_b_screeplot.png" src="../../_images/3_b_screeplot2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plotPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
                <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_c_PCA_reduction&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_c_PCA_reduction2.png"><img alt="3_c_PCA_reduction.png" src="../../_images/3_c_PCA_reduction2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## run UMAP and tSNE on PCA space (default)</span>
<span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_d_UMAP_reduction&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_d_UMAP_reduction2.png"><img alt="3_d_UMAP_reduction.png" src="../../_images/3_d_UMAP_reduction2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">runtSNE</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plotTSNE</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_e_tSNE_reduction&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_e_tSNE_reduction2.png"><img alt="3_e_tSNE_reduction.png" src="../../_images/3_e_tSNE_reduction2.png" style="width: 400px;" /></a>
</div>
<div class="section" id="clustering">
<h2>4. Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## sNN network (default)</span>
<span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">createNearestNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>
<span class="c1">## Leiden clustering</span>
<span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">doLeidenCluster</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">,</span> <span class="n">n_iterations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
        <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;4_a_UMAP_leiden&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/4_a_UMAP_leiden3.png"><img alt="4_a_UMAP_leiden.png" src="../../_images/4_a_UMAP_leiden3.png" style="width: 400px;" /></a>
</div>
<div class="section" id="co-visualize">
<h2>5. Co-Visualize<a class="headerlink" href="#co-visualize" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># expression and spatial</span>
<span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span>
    <span class="n">dim_point_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_a_covis_leiden&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/5_a_covis_leiden2.png"><img alt="5_a_covis_leiden.png" src="../../_images/5_a_covis_leiden2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;nr_genes&#39;</span><span class="p">,</span> <span class="n">color_as_factor</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
    <span class="n">dim_point_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_b_nr_genes&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/5_b_nr_genes1.png"><img alt="5_b_nr_genes.png" src="../../_images/5_b_nr_genes1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="cell-type-marker-gene-detection">
<h2>6. Cell-Type Marker Gene Detection<a class="headerlink" href="#cell-type-marker-gene-detection" title="Permalink to this headline">¶</a></h2>
<div class="section" id="gini">
<h3>6.1 Gini<a class="headerlink" href="#gini" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>gini_markers_subclusters = findMarkers_one_vs_all(gobject = visium_kidney,
                                          method = &#39;gini&#39;,
                                          expression_values = &#39;normalized&#39;,
                                          cluster_column = &#39;leiden_clus&#39;,
                                          min_genes = 20,
                                          min_expr_gini_score = 0.5,
                                          min_det_gini_score = 0.5)
topgenes_gini = gini_markers_subclusters[, head(.SD, 2), by = &#39;cluster&#39;]$genes

# violinplot
violinPlot(visium_kidney, genes = unique(topgenes_gini), cluster_column = &#39;leiden_clus&#39;,
        strip_text = 8, strip_position = &#39;right&#39;,
        save_param = c(save_name = &#39;6_a_violinplot_gini&#39;, base_width = 5, base_height = 10))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_a_violinplot_gini2.png"><img alt="6_a_violinplot_gini.png" src="../../_images/6_a_violinplot_gini2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># cluster heatmap</span>
<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">selected_genes</span> <span class="o">=</span> <span class="n">topgenes_gini</span><span class="p">,</span>
            <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">),</span>
            <span class="n">x_text_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">y_text_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
            <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;6_b_metaheatmap_gini&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_b_metaheatmap_gini2.png"><img alt="6_b_metaheatmap_gini.png" src="../../_images/6_b_metaheatmap_gini2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># umap plots
dimGenePlot2D(visium_kidney, expression_values = &#39;scaled&#39;,
      genes = gini_markers_subclusters[, head(.SD, 1), by = &#39;cluster&#39;]$genes,
      cow_n_col = 3, point_size = 1,
      save_param = c(save_name = &#39;6_c_gini_umap&#39;, base_width = 8, base_height = 5))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_c_gini_umap1.png"><img alt="6_c_gini_umap.png" src="../../_images/6_c_gini_umap1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="scran">
<h3>6.2 Scran<a class="headerlink" href="#scran" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>scran_markers_subclusters = findMarkers_one_vs_all(gobject = visium_kidney,
                                           method = &#39;scran&#39;,
                                           expression_values = &#39;normalized&#39;,
                                           cluster_column = &#39;leiden_clus&#39;)
topgenes_scran = scran_markers_subclusters[, head(.SD, 2), by = &#39;cluster&#39;]$genes

# violinplot
violinPlot(visium_kidney, genes = unique(topgenes_scran), cluster_column = &#39;leiden_clus&#39;,
        strip_text = 10, strip_position = &#39;right&#39;,
        save_param = c(save_name = &#39;6_d_violinplot_scran&#39;, base_width = 5))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_d_violinplot_scran1.png"><img alt="6_d_violinplot_scran.png" src="../../_images/6_d_violinplot_scran1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># cluster heatmap</span>
<span class="n">lotMetaDataHeatmap</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">selected_genes</span> <span class="o">=</span> <span class="n">topgenes_scran</span><span class="p">,</span>
            <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">),</span>
            <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;6_e_metaheatmap_scran&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_e_metaheatmap_scran1.png"><img alt="6_e_metaheatmap_scran.png" src="../../_images/6_e_metaheatmap_scran1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># umap plots
dimGenePlot(visium_kidney, expression_values = &#39;scaled&#39;,
        genes = scran_markers_subclusters[, head(.SD, 1), by = &#39;cluster&#39;]$genes,
        cow_n_col = 3, point_size = 1,
        save_param = c(save_name = &#39;6_f_scran_umap&#39;, base_width = 8, base_height = 5))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_f_scran_umap1.png"><img alt="6_f_scran_umap.png" src="../../_images/6_f_scran_umap1.png" style="width: 400px;" /></a>
</div>
</div>
<div class="section" id="cell-type-annotation">
<h2>7. Cell-Type Annotation<a class="headerlink" href="#cell-type-annotation" title="Permalink to this headline">¶</a></h2>
<p>Visium spatial transcriptomics does not provide single-cell resolution, making cell type annotation a harder problem. Giotto provides 3 ways to calculate enrichment of specific cell-type signature gene list:</p>
<ul class="simple">
<li><p>PAGE</p></li>
<li><p>RANK</p></li>
<li><p>Hypergeometric Test</p></li>
</ul>
<p>See the <a class="reference internal" href="mouse_visium_brain.html#mouse-vis-brain"><span class="std std-ref">Mouse Visium Brain Dataset</span></a> for an example.</p>
</div>
<div class="section" id="spatial-grid">
<h2>8. Spatial Grid<a class="headerlink" href="#spatial-grid" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">visium_kidney</span> <span class="o">&lt;-</span> <span class="n">createSpatialGrid</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
                           <span class="n">sdimx_stepsize</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span>
                           <span class="n">sdimy_stepsize</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span>
                           <span class="n">minimum_padding</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_grid</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
        <span class="n">grid_color</span> <span class="o">=</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">spatial_grid_name</span> <span class="o">=</span> <span class="s1">&#39;spatial_grid&#39;</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_grid&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_grid1.png"><img alt="8_grid.png" src="../../_images/8_grid1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="spatial-network">
<h2>9. Spatial Network<a class="headerlink" href="#spatial-network" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## delaunay network: stats + creation</span>
<span class="n">plotStatDelaunayNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">maximum_distance</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span>
                <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;9_a_delaunay_network&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/9_a_delaunay_network.png"><img alt="9_a_delaunay_network.png" src="../../_images/9_a_delaunay_network.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">visium_kidney</span> <span class="o">=</span> <span class="n">createSpatialNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">minimum_k</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">showNetworks</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">)</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">show_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
        <span class="n">network_color</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">spatial_network_name</span> <span class="o">=</span> <span class="s1">&#39;Delaunay_network&#39;</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;9_b_delaunay_network&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/9_b_delaunay_network.png"><img alt="9_b_delaunay_network.png" src="../../_images/9_b_delaunay_network.png" style="width: 400px;" /></a>
</div>
<div class="section" id="spatial-genes-and-co-expression-patterns">
<h2>10. Spatial Genes and Co-Expression Patterns<a class="headerlink" href="#spatial-genes-and-co-expression-patterns" title="Permalink to this headline">¶</a></h2>
<div class="section" id="spatial-genes">
<h3>10.1 Spatial Genes<a class="headerlink" href="#spatial-genes" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## kmeans binarization
kmtest = binSpect(visium_kidney)
spatGenePlot(visium_kidney, expression_values = &#39;scaled&#39;,
     genes = kmtest$genes[1:6], cow_n_col = 2, point_size = 1.5,
     save_param = c(save_name = &#39;10_a_spatial_genes_km&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_a_spatial_genes_km1.png"><img alt="10_a_spatial_genes_km.png" src="../../_images/10_a_spatial_genes_km1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## rank binarization
ranktest = binSpect(visium_kidney, bin_method = &#39;rank&#39;)
spatGenePlot(visium_kidney, expression_values = &#39;scaled&#39;,
     genes = ranktest$genes[1:6], cow_n_col = 2, point_size = 1.5,
     save_param = c(save_name = &#39;10_b_spatial_genes_rank&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_b_spatial_genes_rank1.png"><img alt="10_b_spatial_genes_rank.png" src="../../_images/10_b_spatial_genes_rank1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="spatial-co-expression-patterns">
<h3>10.2 Spatial Co-Expression Patterns<a class="headerlink" href="#spatial-co-expression-patterns" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## spatially correlated genes ##
ext_spatial_genes = kmtest[1:500]$genes

# 1. calculate gene spatial correlation and single-cell correlation
# create spatial correlation object
spat_cor_netw_DT = detectSpatialCorGenes(visium_kidney,
                                 method = &#39;network&#39;,
                                 spatial_network_name = &#39;Delaunay_network&#39;,
                                 subset_genes = ext_spatial_genes)

# 2. identify most similar spatially correlated genes for one gene
Napsa_top10_genes = showSpatialCorGenes(spat_cor_netw_DT, genes = &#39;Napsa&#39;, show_top_genes = 10)
spatGenePlot(visium_kidney, expression_values = &#39;scaled&#39;,
     genes = c(&#39;Napsa&#39;, &#39;Kap&#39;, &#39;Defb29&#39;, &#39;Prdx1&#39;), point_size = 3,
     save_param = c(save_name = &#39;10_d_Napsa_correlated_genes&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_d_Napsa_correlated_genes.png"><img alt="10_d_Napsa_correlated_genes.png" src="../../_images/10_d_Napsa_correlated_genes.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># 3. cluster correlated genes &amp; visualize</span>
<span class="n">spat_cor_netw_DT</span> <span class="o">=</span> <span class="n">clusterSpatialCorGenes</span><span class="p">(</span><span class="n">spat_cor_netw_DT</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;spat_netw_clus&#39;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">8</span><span class="p">)</span>

<span class="n">heatmSpatialCorGenes</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">spatCorObject</span> <span class="o">=</span> <span class="n">spat_cor_netw_DT</span><span class="p">,</span> <span class="n">use_clus_name</span> <span class="o">=</span> <span class="s1">&#39;spat_netw_clus&#39;</span><span class="p">,</span>
             <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;10_e_heatmap_correlated_genes&#39;</span><span class="p">,</span> <span class="n">save_format</span> <span class="o">=</span> <span class="s1">&#39;pdf&#39;</span><span class="p">,</span>
                            <span class="n">base_height</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="n">base_width</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">units</span> <span class="o">=</span> <span class="s1">&#39;cm&#39;</span><span class="p">),</span>
             <span class="n">heatmap_legend_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">title</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_e_heatmap_correlated_genes.png"><img alt="10_e_heatmap_correlated_genes.png" src="../../_images/10_e_heatmap_correlated_genes.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># 4. rank spatial correlated clusters and show genes for selected clusters</span>
<span class="n">netw_ranks</span> <span class="o">=</span> <span class="n">rankSpatialCorGroups</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">,</span> <span class="n">spatCorObject</span> <span class="o">=</span> <span class="n">spat_cor_netw_DT</span><span class="p">,</span> <span class="n">use_clus_name</span> <span class="o">=</span> <span class="s1">&#39;spat_netw_clus&#39;</span><span class="p">,</span>
                          <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;10_f_rank_correlated_groups&#39;</span><span class="p">,</span>
                                         <span class="n">base_height</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">base_width</span> <span class="o">=</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_f_rank_correlated_groups.png"><img alt="10_f_rank_correlated_groups.png" src="../../_images/10_f_rank_correlated_groups.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>top_netw_spat_cluster = showSpatialCorGenes(spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;,
                                    selected_clusters = 6, show_top_genes = 1)

# 5. create metagene enrichment score for clusters
cluster_genes_DT = showSpatialCorGenes(spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;, show_top_genes = 1)
cluster_genes = cluster_genes_DT$clus; names(cluster_genes) = cluster_genes_DT$gene_ID

visium_kidney = createMetagenes(visium_kidney, gene_clusters = cluster_genes, name = &#39;cluster_metagene&#39;)

spatCellPlot(visium_kidney,
     spat_enr_names = &#39;cluster_metagene&#39;,
     cell_annotation_values = netw_ranks$clusters,
     point_size = 1.5, cow_n_col = 4,
     save_param = c(save_name = &#39;10_g_spat_enrichment_score_plots&#39;,
                    base_width = 13, base_height = 6))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_g_spat_enrichment_score_plots.png"><img alt="10_g_spat_enrichment_score_plots.png" src="../../_images/10_g_spat_enrichment_score_plots.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># example for gene per cluster
top_netw_spat_cluster = showSpatialCorGenes(spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;,
                                    selected_clusters = 1:8, show_top_genes = 1)
first_genes = top_netw_spat_cluster[, head(.SD, 1), by = clus]$gene_ID
cluster_names = top_netw_spat_cluster[, head(.SD, 1), by = clus]$clus
names(first_genes) = cluster_names
first_genes = first_genes[as.character(netw_ranks$clusters)]

spatGenePlot(visium_kidney, genes = first_genes, expression_values = &#39;scaled&#39;, cow_n_col = 4, midpoint = 0, point_size = 2,
     save_param = c(save_name = &#39;10_h_spat_enrichment_score_plots_genes&#39;,
                    base_width = 11, base_height = 6))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_h_spat_enrichment_score_plots_genes.png"><img alt="10_h_spat_enrichment_score_plots_genes.png" src="../../_images/10_h_spat_enrichment_score_plots_genes.png" style="width: 400px;" /></a>
</div>
</div>
<div class="section" id="hmrf-domains">
<h2>11. HMRF Domains<a class="headerlink" href="#hmrf-domains" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># HMRF requires a fully connected network!
visium_kidney = createSpatialNetwork(gobject = visium_kidney, minimum_k = 2, name = &#39;Delaunay_full&#39;)

# spatial genes
my_spatial_genes &lt;- kmtest[1:100]$genes

# do HMRF with different betas
hmrf_folder = paste0(results_folder,&#39;/&#39;,&#39;11_HMRF/&#39;)
if(!file.exists(hmrf_folder)) dir.create(hmrf_folder, recursive = T)

HMRF_spatial_genes = doHMRF(gobject = visium_kidney, expression_values = &#39;scaled&#39;,
                    spatial_network_name = &#39;Delaunay_full&#39;,
                    spatial_genes = my_spatial_genes,
                    k = 5,
                    betas = c(0, 1, 6),
                    output_folder = paste0(hmrf_folder, &#39;/&#39;, &#39;Spatial_genes/SG_topgenes_k5_scaled&#39;))

## view results of HMRF
for(i in seq(0, 5, by = 1)) {
        viewHMRFresults2D(gobject = visium_kidney,
            HMRFoutput = HMRF_spatial_genes,
            k = 5, betas_to_view = i,
            point_size = 2)
}
</pre></div>
</div>
<p><strong>Alternative Way to View Results</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#results = writeHMRFresults(gobject = ST_test,</span>
<span class="c1">#                           HMRFoutput = HMRF_spatial_genes,</span>
<span class="c1">#                           k = 5, betas_to_view = seq(0, 25, by = 5))</span>
<span class="c1">#ST_test = addCellMetadata(ST_test, new_metadata = results, by_column = T, column_cell_ID = &#39;cell_ID&#39;)</span>


<span class="c1">## add HMRF of interest to giotto object</span>
<span class="n">visium_kidney</span> <span class="o">=</span> <span class="n">addHMRF</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
                <span class="n">HMRFoutput</span> <span class="o">=</span> <span class="n">HMRF_spatial_genes</span><span class="p">,</span>
                <span class="n">k</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">betas_to_add</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">hmrf_name</span> <span class="o">=</span> <span class="s1">&#39;HMRF&#39;</span><span class="p">)</span>

<span class="c1">## visualize</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;HMRF_k5_b.0&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;11_a_HMRF_k5_b.0&#39;</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;HMRF_k5_b.2&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;11_b_HMRF_k5_b.2&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="section" id="export-and-create-giotto-viewer">
<h2>Export and Create Giotto Viewer<a class="headerlink" href="#export-and-create-giotto-viewer" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check which annotations are available</span>
<span class="n">combineMetadata</span><span class="p">(</span><span class="n">visium_kidney</span><span class="p">)</span>

<span class="c1"># select annotations, reductions and expression values to view in Giotto Viewer</span>
<span class="n">viewer_folder</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">results_folder</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_visium_kidney_viewer&#39;</span><span class="p">)</span>

<span class="n">exportGiottoViewer</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_kidney</span><span class="p">,</span>
           <span class="n">output_directory</span> <span class="o">=</span> <span class="n">viewer_folder</span><span class="p">,</span>
           <span class="n">spat_enr_names</span> <span class="o">=</span> <span class="s1">&#39;PAGE&#39;</span><span class="p">,</span>
           <span class="n">factor_annotations</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;leiden_clus&#39;</span><span class="p">),</span>
           <span class="n">numeric_annotations</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;nr_genes&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;clus_25&#39;</span><span class="p">),</span>
           <span class="n">dim_reductions</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span><span class="p">),</span>
           <span class="n">dim_reduction_names</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span><span class="p">),</span>
           <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;scaled&#39;</span><span class="p">,</span>
           <span class="n">expression_rounding</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
           <span class="n">overwrite_dir</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Ruben Dries, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>