<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Mouse Visium Brain &mdash; Giotto 1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/sphinx-toolbox-code.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/sphinx-toolbox.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/tabs.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/sphinx_toolbox_installation.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
        <script src="../../_static/sphinx_toolbox_installation.js"></script>
        <script src="../../_static/design-tabs.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Mouse Visium Kidney" href="mouse_visium_kidney.html" />
    <link rel="prev" title="STARmap Mouse Cortex" href="STARmap_mouse_cortex.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html">
            <img src="../../_static/GiottoLogo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../introduction.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Development</a></li>
<li class="toctree-l1"><a class="reference external" href="http://spatial.rc.fas.harvard.edu/giotto-viewer/">Giotto Viewer</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Guides</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../tipsandtricks.html">HowTos</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../datasets.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../datasets.html#dataset-information">Dataset Information</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../datasets.html#example-datasets">Example Datasets</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../datasets.html#mini-datasets">Mini Datasets</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../../datasets.html#full-datasets">Full Datasets</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="seqFISH_cortex.html">    Mouse seqFISH Cortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="merFISH_hypot_preopt_region.html">Mouse merFISH Hypoth. Preopt. Region</a></li>
<li class="toctree-l4"><a class="reference internal" href="STARmap_mouse_cortex.html">Mouse STARmap Cortex</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">    Mouse Visium Brain</a></li>
<li class="toctree-l4"><a class="reference internal" href="mouse_visium_kidney.html">    Mouse Visium Kidney</a></li>
<li class="toctree-l4"><a class="reference internal" href="mouse_CODEX_spleen.html">    Mouse CODEX Spleen</a></li>
<li class="toctree-l4"><a class="reference internal" href="osmFISH_mouse_SS_cortex.html">    Mouse osmFISH SScortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="human_CyCIF_PDAC.html">    Human CyCIF PDAC</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../trygiotto.html">Try Giotto</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Issues</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../errorsfaqsandtips.html">Common Errors and Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../github_issues.html">Report a Bug</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Giotto</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../../datasets.html">Examples</a> &raquo;</li>
      <li>Mouse Visium Brain</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/subsections/datasets/mouse_visium_brain.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="mouse-visium-brain">
<span id="id1"></span><h1>Mouse Visium Brain<a class="headerlink" href="#mouse-visium-brain" title="Permalink to this headline"></a></h1>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>You’re using the same Giotto version with which this tutorial was written.</p>
</div>
<div class="section" id="install-python-modules">
<h2>Install Python Modules<a class="headerlink" href="#install-python-modules" title="Permalink to this headline"></a></h2>
<p>To run this vignette you need to install <strong>all</strong> of the necessary Python modules.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Python module installation can be done either <strong>automatically</strong> via our installation tool (from within R) (see step 2.2A) or <strong>manually</strong> (see step 2.2B).</p>
<p><svg version="1.1" width="1.0em" height="1.0em" class="sd-octicon sd-octicon-eye" viewBox="0 0 16 16" aria-hidden="true"><path fill-rule="evenodd" d="M1.679 7.932c.412-.621 1.242-1.75 2.366-2.717C5.175 4.242 6.527 3.5 8 3.5c1.473 0 2.824.742 3.955 1.715 1.124.967 1.954 2.096 2.366 2.717a.119.119 0 010 .136c-.412.621-1.242 1.75-2.366 2.717C10.825 11.758 9.473 12.5 8 12.5c-1.473 0-2.824-.742-3.955-1.715C2.92 9.818 2.09 8.69 1.679 8.068a.119.119 0 010-.136zM8 2c-1.981 0-3.67.992-4.933 2.078C1.797 5.169.88 6.423.43 7.1a1.619 1.619 0 000 1.798c.45.678 1.367 1.932 2.637 3.024C4.329 13.008 6.019 14 8 14c1.981 0 3.67-.992 4.933-2.078 1.27-1.091 2.187-2.345 2.637-3.023a1.619 1.619 0 000-1.798c-.45-.678-1.367-1.932-2.637-3.023C11.671 2.992 9.981 2 8 2zm0 8a2 2 0 100-4 2 2 0 000 4z"></path></svg> See <a class="reference internal" href="../../gettingstarted.html#part2-python-giotto-requirements"><span class="std std-ref">Part 2.2 Giotto-Specific Python Packages</span></a> of our Giotto Installation section for step-by-step instructions.</p>
</div>
</div>
<div class="section" id="set-up-giotto">
<h2>Set-Up Giotto<a class="headerlink" href="#set-up-giotto" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Giotto</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="set-a-working-directory">
<h3>Set A Working Directory<a class="headerlink" href="#set-a-working-directory" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#results_folder = &#39;/path/to/directory/&#39;</span>
<span class="n">results_folder</span> <span class="o">=</span> <span class="s1">&#39;/Volumes/Ruben_Seagate/Dropbox (Personal)/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Results/Visium/Brain/201226_results//&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="set-a-giotto-python-path">
<h3>Set A Giotto Python Path<a class="headerlink" href="#set-a-giotto-python-path" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># set python path to your preferred python version path</span>
<span class="c1"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span>
<span class="n">python_path</span> <span class="o">=</span> <span class="n">NULL</span>
<span class="k">if</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">python_path</span><span class="p">))</span> <span class="p">{</span>
    <span class="n">installGiottoEnvironment</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="dataset-explanation">
<h2>Dataset Explanation<a class="headerlink" href="#dataset-explanation" title="Permalink to this headline"></a></h2>
<p><a class="reference external" href="https://www.10xgenomics.com/spatial-transcriptomics">10X genomics</a> recently launched a new platform to obtain spatial expression data using a Visium Spatial Gene Expression slide. The Visium brain data to run this tutorial can be found <a class="reference external" href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Adult_Mouse_Brain">here</a>.</p>
<div class="dropdown admonition note">
<p class="admonition-title">Note</p>
<p>Visium Technology</p>
<a class="reference internal image-reference" href="../../_images/visium_technology.png"><img alt="Visium Technology" src="../../_images/visium_technology.png" style="width: 500px;" /></a>
</div>
<p>High resolution png from original tissue:</p>
<a class="reference internal image-reference" href="../../_images/visium_brain_image_summary1.png"><img alt="visium_brain_image_summary.png" src="../../_images/visium_brain_image_summary1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="dataset-download">
<h2>Dataset Download<a class="headerlink" href="#dataset-download" title="Permalink to this headline"></a></h2>
<p>The merFISH data to run this tutorial can be found <a class="reference external" href="https://github.com/RubD/spatial-datasets/tree/master/data/2018_merFISH_science_hypo_preoptic">here</a>. Alternatively you can use the <strong>getSpatialDataset</strong> to automatically download this dataset like we do in this example.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># download data to working directory ####</span>
<span class="c1"># if wget is installed, set method = &#39;wget&#39;</span>
<span class="c1"># if you run into authentication issues with wget, then add &quot; extra = &#39;--no-check-certificate&#39; &quot;</span>
<span class="n">getSpatialDataset</span><span class="p">(</span><span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;cycif_PDAC&#39;</span><span class="p">,</span> <span class="n">directory</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;wget&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="giotto-global-instructions-and-preparations">
<h2>1. Giotto Global Instructions and Preparations<a class="headerlink" href="#giotto-global-instructions-and-preparations" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## create instructions</span>
     <span class="n">instrs</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span>
                   <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                   <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>

     <span class="c1">## provide path to visium folder</span>
     <span class="c1">#data_path = &#39;/path/to/Brain_data/&#39;</span>
     <span class="n">data_path</span> <span class="o">=</span> <span class="s1">&#39;/Volumes/Ruben_Seagate/Dropbox (Personal)/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Data/Visium_data/Brain_data/&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="create-giotto-object-and-process-the-data">
<h2>2. Create Giotto Object and Process The Data<a class="headerlink" href="#create-giotto-object-and-process-the-data" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## directly from visium folder</span>
<span class="n">visium_brain</span> <span class="o">=</span> <span class="n">createGiottoVisiumObject</span><span class="p">(</span><span class="n">visium_dir</span> <span class="o">=</span> <span class="n">data_path</span><span class="p">,</span> <span class="n">expr_data</span> <span class="o">=</span> <span class="s1">&#39;raw&#39;</span><span class="p">,</span>
                 <span class="n">png_name</span> <span class="o">=</span> <span class="s1">&#39;tissue_lowres_image.png&#39;</span><span class="p">,</span>
                 <span class="n">gene_column_index</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">instructions</span> <span class="o">=</span> <span class="n">instrs</span><span class="p">)</span>

<span class="c1">## update and align background image</span>
<span class="c1"># problem: image is not perfectly aligned</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span> <span class="n">show_image</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_a_spatplot_image&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_a_spatplot_image.png"><img alt="2_a_spatplot_image.png" src="../../_images/2_a_spatplot_image.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check name</span>
<span class="n">showGiottoImageNames</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">)</span> <span class="c1"># &quot;image&quot; is the default name</span>
<span class="c1"># adjust parameters to align image (iterative approach)</span>
<span class="n">visium_brain</span> <span class="o">=</span> <span class="n">updateGiottoImage</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">image_name</span> <span class="o">=</span> <span class="s1">&#39;image&#39;</span><span class="p">,</span>
              <span class="n">xmax_adj</span> <span class="o">=</span> <span class="mi">1300</span><span class="p">,</span> <span class="n">xmin_adj</span> <span class="o">=</span> <span class="mi">1200</span><span class="p">,</span>
              <span class="n">ymax_adj</span> <span class="o">=</span> <span class="mi">1100</span><span class="p">,</span> <span class="n">ymin_adj</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>

<span class="c1"># now it&#39;s aligned</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span> <span class="n">show_image</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_b_spatplot_image_adjusted&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_b_spatplot_image_adjusted.png"><img alt="2_b_spatplot_image_adjusted.png" src="../../_images/2_b_spatplot_image_adjusted.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## check metadata</span>
<span class="n">pDataDT</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">)</span>

<span class="c1">## compare in tissue with provided jpg</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;0&#39;</span> <span class="o">=</span> <span class="s1">&#39;lightgrey&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span><span class="p">),</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_c_in_tissue&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_c_in_tissue.png"><img alt="2_c_in_tissue.png" src="../../_images/2_c_in_tissue.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## subset on spots that were covered by tissue
metadata = pDataDT(visium_brain)
in_tissue_barcodes = metadata[in_tissue == 1]$cell_ID
visium_brain = subsetGiotto(visium_brain, cell_ids = in_tissue_barcodes)

## filter
visium_brain &lt;- filterGiotto(gobject = visium_brain,
              expression_threshold = 1,
              gene_det_in_min_cells = 50,
              min_det_genes_per_cell = 1000,
              expression_values = c(&#39;raw&#39;),
              verbose = T)

## normalize
visium_brain &lt;- normalizeGiotto(gobject = visium_brain, scalefactor = 6000, verbose = T)

## add gene &amp; cell statistics
visium_brain &lt;- addStatistics(gobject = visium_brain)

## visualize
patPlot2D(gobject = visium_brain, show_image = T, point_alpha = 0.7,
    ave_param = list(save_name = &#39;2_d_spatial_locations&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_d_spatial_locations1.png"><img alt="2_d_spatial_locations.png" src="../../_images/2_d_spatial_locations1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">show_image</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
    <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;nr_genes&#39;</span><span class="p">,</span> <span class="n">color_as_factor</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_e_nr_genes&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_e_nr_genes.png"><img alt="2_e_nr_genes.png" src="../../_images/2_e_nr_genes.png" style="width: 400px;" /></a>
</div>
<div class="section" id="dimension-reduction">
<h2>3. Dimension Reduction<a class="headerlink" href="#dimension-reduction" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## highly variable genes (HVG)</span>
<span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">calculateHVG</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
              <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_a_HVGplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_a_HVGplot.png"><img alt="3_a_HVGplot.png" src="../../_images/3_a_HVGplot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## run PCA on expression values (default)
gene_metadata = fDataDT(visium_brain)
featgenes = gene_metadata[hvg == &#39;yes&#39; &amp; perc_cells &gt; 3 &amp; mean_expr_det &gt; 0.4]$gene_ID

visium_brain &lt;- runPCA(gobject = visium_brain,
           genes_to_use = featgenes,
           scale_unit = F, center = T,
           method=&quot;factominer&quot;)

screePlot(visium_brain, ncp = 30, save_param = list(save_name = &#39;3_b_screeplot&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_b_screeplot.png"><img alt="3_b_screeplot.png" src="../../_images/3_b_screeplot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plotPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_c_PCA_reduction&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_c_PCA_reduction.png"><img alt="3_c_PCA_reduction.png" src="../../_images/3_c_PCA_reduction.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## run UMAP and tSNE on PCA space (default)</span>
<span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_d_UMAP_reduction&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_d_UMAP_reduction.png"><img alt="3_d_UMAP_reduction.png" src="../../_images/3_d_UMAP_reduction.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">runtSNE</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plotTSNE</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_e_tSNE_reduction&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_e_tSNE_reduction.png"><img alt="3_e_tSNE_reduction.png" src="../../_images/3_e_tSNE_reduction.png" style="width: 400px;" /></a>
</div>
<div class="section" id="clustering">
<h2>4. Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## sNN network (default)</span>
<span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">createNearestNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>
<span class="c1">## Leiden clustering</span>
<span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">doLeidenCluster</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">,</span> <span class="n">n_iterations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
    <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;4_a_UMAP_leiden&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/4_a_UMAP_leiden1.png"><img alt="4_a_UMAP_leiden.png" src="../../_images/4_a_UMAP_leiden1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="whole-dataset">
<h2>5. Whole Dataset<a class="headerlink" href="#whole-dataset" title="Permalink to this headline"></a></h2>
<div class="section" id="expression-and-spatial">
<h3>5.1 Expression and Spatial<a class="headerlink" href="#expression-and-spatial" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># expression and spatial</span>
<span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span>
    <span class="n">dim_point_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_a_covis_leiden&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/5_a_covis_leiden.png"><img alt="5_a_covis_leiden.png" src="../../_images/5_a_covis_leiden.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;nr_genes&#39;</span><span class="p">,</span> <span class="n">color_as_factor</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
    <span class="n">dim_point_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_b_nr_genes&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/5_b_nr_genes.png"><img alt="5_b_nr_genes.png" src="../../_images/5_b_nr_genes.png" style="width: 400px;" /></a>
</div>
<div class="section" id="subset-dataset-on-dg-region">
<h3>5.2 Subset Dataset on DG Region<a class="headerlink" href="#subset-dataset-on-dg-region" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">DG_subset</span> <span class="o">=</span> <span class="n">subsetGiottoLocs</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span>
             <span class="n">x_max</span> <span class="o">=</span> <span class="mi">6500</span><span class="p">,</span> <span class="n">x_min</span> <span class="o">=</span> <span class="mi">3000</span><span class="p">,</span>
             <span class="n">y_max</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2500</span><span class="p">,</span> <span class="n">y_min</span> <span class="o">=</span> <span class="o">-</span><span class="mi">5500</span><span class="p">,</span>
             <span class="n">return_gobject</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>

<span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">DG_subset</span><span class="p">,</span>
    <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
     <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_c_DEG_subset&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/5_c_DEG_subset.png"><img alt="5_c_DEG_subset.png" src="../../_images/5_c_DEG_subset.png" style="width: 400px;" /></a>
</div>
</div>
<div class="section" id="cell-type-marker-gene-detection">
<h2>6. Cell-Type Marker Gene Detection<a class="headerlink" href="#cell-type-marker-gene-detection" title="Permalink to this headline"></a></h2>
<div class="section" id="gini">
<h3>6.1 Gini<a class="headerlink" href="#gini" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>gini_markers_subclusters = findMarkers_one_vs_all(gobject = visium_brain,
                      method = &#39;gini&#39;,
                      expression_values = &#39;normalized&#39;,
                      cluster_column = &#39;leiden_clus&#39;,
                      min_genes = 20,
                      min_expr_gini_score = 0.5,
                      min_det_gini_score = 0.5)
topgenes_gini = gini_markers_subclusters[, head(.SD, 2), by = &#39;cluster&#39;]$genes

# violinplot
violinPlot(visium_brain, genes = unique(topgenes_gini), cluster_column = &#39;leiden_clus&#39;,
    strip_text = 8, strip_position = &#39;right&#39;,
    save_param = list(save_name = &#39;6_a_violinplot_gini&#39;, base_width = 5, base_height = 10))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_a_violinplot_gini.png"><img alt="6_a_violinplot_gini.png" src="../../_images/6_a_violinplot_gini.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># cluster heatmap</span>
<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">selected_genes</span> <span class="o">=</span> <span class="n">topgenes_gini</span><span class="p">,</span>
        <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">),</span>
        <span class="n">x_text_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">y_text_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;6_b_metaheatmap_gini&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_b_metaheatmap_gini.png"><img alt="6_b_metaheatmap_gini.png" src="../../_images/6_b_metaheatmap_gini.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># umap plots
dimGenePlot2D(visium_brain, expression_values = &#39;scaled&#39;,
      genes = gini_markers_subclusters[, head(.SD, 1), by = &#39;cluster&#39;]$genes,
      cow_n_col = 3, point_size = 1,
      save_param = list(save_name = &#39;6_c_gini_umap&#39;, base_width = 8, base_height = 5))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_c_gini_umap.png"><img alt="6_c_gini_umap.png" src="../../_images/6_c_gini_umap.png" style="width: 400px;" /></a>
</div>
<div class="section" id="scran">
<h3>6.2 Scran<a class="headerlink" href="#scran" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>scran_markers_subclusters = findMarkers_one_vs_all(gobject = visium_brain,
                       method = &#39;scran&#39;,
                       expression_values = &#39;normalized&#39;,
                       cluster_column = &#39;leiden_clus&#39;)
topgenes_scran = scran_markers_subclusters[, head(.SD, 2), by = &#39;cluster&#39;]$genes

# violinplot
violinPlot(visium_brain, genes = unique(topgenes_scran), cluster_column = &#39;leiden_clus&#39;,
    strip_text = 10, strip_position = &#39;right&#39;,
    save_param = list(save_name = &#39;6_d_violinplot_scran&#39;, base_width = 5))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_d_violinplot_scran.png"><img alt="6_d_violinplot_scran.png" src="../../_images/6_d_violinplot_scran.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># cluster heatmap</span>
<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">selected_genes</span> <span class="o">=</span> <span class="n">topgenes_scran</span><span class="p">,</span>
        <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">),</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_e_metaheatmap_scran.png"><img alt="6_e_metaheatmap_scran.png" src="../../_images/6_e_metaheatmap_scran.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># umap plots
dimGenePlot(visium_brain, expression_values = &#39;scaled&#39;,
    genes = scran_markers_subclusters[, head(.SD, 1), by = &#39;cluster&#39;]$genes,
    cow_n_col = 3, point_size = 1,
    save_param = list(save_name = &#39;6_f_scran_umap&#39;, base_width = 8, base_height = 5))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_f_scran_umap.png"><img alt="6_f_scran_umap.png" src="../../_images/6_f_scran_umap.png" style="width: 400px;" /></a>
</div>
</div>
<div class="section" id="cell-type-annotation">
<span id="mouse-vis-brain"></span><h2>7. Cell-Type Annotation<a class="headerlink" href="#cell-type-annotation" title="Permalink to this headline"></a></h2>
<p>Visium spatial transcriptomics does not provide single-cell resolution, making cell type annotation a harder problem. Giotto provides 3 ways to calculate enrichment of specific cell-type signature gene list:</p>
<ul class="simple">
<li><p>PAGE</p></li>
<li><p>RANK</p></li>
<li><p>Hypergeometric Test</p></li>
</ul>
<p>Known markers for different mouse brain cell types: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6086934/">Zeisel, A. et al. Molecular Architecture of the Mouse Nervous System. Cell 174, 999-1014.e22 (2018)</a> .
Cell type signatures are combination of all marker genes identified in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6086934/">Zeisel et al.</a></p>
<div class="section" id="page-enrichment">
<h3>7.1 PAGE Enrichment<a class="headerlink" href="#page-enrichment" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># 1.1 create binary matrix of cell signature genes
# small example #
gran_markers = c(&quot;Nr3c2&quot;, &quot;Gabra5&quot;, &quot;Tubgcp2&quot;, &quot;Ahcyl2&quot;,
     &quot;Islr2&quot;, &quot;Rasl10a&quot;, &quot;Tmem114&quot;, &quot;Bhlhe22&quot;,
     &quot;Ntf3&quot;, &quot;C1ql2&quot;)

oligo_markers = c(&quot;Efhd1&quot;, &quot;H2-Ab1&quot;, &quot;Enpp6&quot;, &quot;Ninj2&quot;,
      &quot;Bmp4&quot;, &quot;Tnr&quot;, &quot;Hapln2&quot;, &quot;Neu4&quot;,
      &quot;Wfdc18&quot;, &quot;Ccp110&quot;)

di_mesench_markers = c(&quot;Cartpt&quot;, &quot;Scn1a&quot;, &quot;Lypd6b&quot;,  &quot;Drd5&quot;,
           &quot;Gpr88&quot;, &quot;Plcxd2&quot;, &quot;Cpne7&quot;, &quot;Pou4f1&quot;,
           &quot;Ctxn2&quot;, &quot;Wnt4&quot;)

signature_matrix = makeSignMatrixPAGE(sign_names = c(&#39;Granule_neurons&#39;,
                         &#39;Oligo_dendrocytes&#39;,
                         &#39;di_mesenchephalon&#39;),
                  sign_list = list(gran_markers,
                           oligo_markers,
                           di_mesench_markers))

# 1.2 [shortcut] fully pre-prepared matrix for all cell types
sign_matrix_path = system.file(&quot;extdata&quot;, &quot;sig_matrix.txt&quot;, package = &#39;Giotto&#39;)
brain_sc_markers = data.table::fread(sign_matrix_path)
sig_matrix = as.matrix(brain_sc_markers[,-1]); rownames(sig_matrix) = brain_sc_markers$Event



# 1.3 enrichment test with PAGE

# runSpatialEnrich() can also be used as a wrapper for all currently provided enrichment options
visium_brain = runPAGEEnrich(gobject = visium_brain, sign_matrix = sig_matrix)


# 1.4 heatmap of enrichment versus annotation (e.g. clustering result)
cell_types = colnames(sig_matrix)
plotMetaDataCellsHeatmap(gobject = visium_brain,
         metadata_cols = &#39;leiden_clus&#39;,
         value_cols = cell_types,
         spat_enr_names = &#39;PAGE&#39;,
         x_text_size = 8,
         y_text_size = 8,
         save_param = list(save_name=&quot;7_a_metaheatmap&quot;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_a_metaheatmap2.png"><img alt="7_a_metaheatmap.png" src="../../_images/7_a_metaheatmap2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1.5 visualizations</span>
<span class="n">cell_types_subset</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">sig_matrix</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
<span class="n">spatCellPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
     <span class="n">spat_enr_names</span> <span class="o">=</span> <span class="s1">&#39;PAGE&#39;</span><span class="p">,</span>
     <span class="n">cell_annotation_values</span> <span class="o">=</span> <span class="n">cell_types_subset</span><span class="p">,</span>
     <span class="n">cow_n_col</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span><span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span>
     <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span><span class="o">=</span><span class="s2">&quot;7_b_spatcellplot_1&quot;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_b_spatcellplot.png"><img alt="7_b_spatcellplot_1.png" src="../../_images/7_b_spatcellplot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cell_types_subset</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">sig_matrix</span><span class="p">)[</span><span class="mi">11</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span>
<span class="n">spatCellPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">spat_enr_names</span> <span class="o">=</span> <span class="s1">&#39;PAGE&#39;</span><span class="p">,</span>
     <span class="n">cell_annotation_values</span> <span class="o">=</span> <span class="n">cell_types_subset</span><span class="p">,</span> <span class="n">cow_n_col</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
     <span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span>
     <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span><span class="o">=</span><span class="s2">&quot;7_c_spatcellplot_2&quot;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_c_spatcellplot.png"><img alt="7_c_spatcellplot_2.png" src="../../_images/7_c_spatcellplot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatDimCellPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
    <span class="n">spat_enr_names</span> <span class="o">=</span> <span class="s1">&#39;PAGE&#39;</span><span class="p">,</span>
    <span class="n">cell_annotation_values</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;Cortex_hippocampus&#39;</span><span class="p">,</span> <span class="s1">&#39;Granule_neurons&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;di_mesencephalon_1&#39;</span><span class="p">,</span> <span class="s1">&#39;Oligo_dendrocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Vascular&#39;</span><span class="p">),</span>
    <span class="n">cow_n_col</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">plot_alignment</span> <span class="o">=</span> <span class="s1">&#39;horizontal&#39;</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span><span class="o">=</span><span class="s2">&quot;7_d_spatDimCellPlot&quot;</span><span class="p">,</span> <span class="n">base_width</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">base_height</span><span class="o">=</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_d_spatDimCellPlot.png"><img alt="7_d_spatDimCellPlot.png" src="../../_images/7_d_spatDimCellPlot.png" style="width: 400px;" /></a>
</div>
</div>
<div class="section" id="spatial-grid">
<h2>8. Spatial Grid<a class="headerlink" href="#spatial-grid" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">createSpatialGrid</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
               <span class="n">sdimx_stepsize</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span>
               <span class="n">sdimy_stepsize</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span>
               <span class="n">minimum_padding</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_grid</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
    <span class="n">grid_color</span> <span class="o">=</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">spatial_grid_name</span> <span class="o">=</span> <span class="s1">&#39;spatial_grid&#39;</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_grid&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_grid.png"><img alt="8_grid.png" src="../../_images/8_grid.png" style="width: 400px;" /></a>
</div>
<div class="section" id="spatial-network">
<h2>9. Spatial Network<a class="headerlink" href="#spatial-network" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">visium_brain</span> <span class="o">&lt;-</span> <span class="n">createSpatialNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
                 <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;kNN&#39;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
                 <span class="n">maximum_distance_knn</span> <span class="o">=</span> <span class="mi">400</span><span class="p">,</span>
                 <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;spatial_network&#39;</span><span class="p">)</span>

<span class="n">showNetworks</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">)</span>

<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span> <span class="n">show_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
    <span class="n">network_color</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">spatial_network_name</span> <span class="o">=</span> <span class="s1">&#39;spatial_network&#39;</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;9_a_knn_network&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/9_a_knn_network.png"><img alt="9_a_knn_network.png" src="../../_images/9_a_knn_network.png" style="width: 400px;" /></a>
</div>
<div class="section" id="spatial-genes-and-patterns">
<h2>10. Spatial Genes and Patterns<a class="headerlink" href="#spatial-genes-and-patterns" title="Permalink to this headline"></a></h2>
<div class="section" id="spatial-genes">
<h3>10.1 Spatial Genes<a class="headerlink" href="#spatial-genes" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## kmeans binarization
kmtest = binSpect(visium_brain, calc_hub = T, hub_min_int = 5,
      spatial_network_name = &#39;spatial_network&#39;)
spatGenePlot(visium_brain, expression_values = &#39;scaled&#39;,
     genes = kmtest$genes[1:6], cow_n_col = 2, point_size = 1.5,
     save_param = list(save_name = &#39;10_a_spatial_genes_km&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_a_spatial_genes_km.png"><img alt="10_a_spatial_genes_km.png" src="../../_images/10_a_spatial_genes_km.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## rank binarization
ranktest = binSpect(visium_brain, bin_method = &#39;rank&#39;,
        calc_hub = T, hub_min_int = 5,
        spatial_network_name = &#39;spatial_network&#39;)
spatGenePlot(visium_brain, expression_values = &#39;scaled&#39;,
     genes = ranktest$genes[1:6], cow_n_col = 2, point_size = 1.5,
     save_param = list(save_name = &#39;10_b_spatial_genes_rank&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_b_spatial_genes_rank.png"><img alt="10_b_spatial_genes_rank.png" src="../../_images/10_b_spatial_genes_rank.png" style="width: 400px;" /></a>
</div>
<div class="section" id="spatial-patterns">
<h3>10.2 Spatial Patterns<a class="headerlink" href="#spatial-patterns" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># cluster the top 1500 spatial genes into 20 clusters
ext_spatial_genes = ranktest[1:1500,]$gene

# here we use existing detectSpatialCorGenes function to calculate pairwise distances between genes (but set network_smoothing=0 to use default clustering)
spat_cor_netw_DT = detectSpatialCorGenes(visium_brain,
                 method = &#39;network&#39;,
                 spatial_network_name = &#39;spatial_network&#39;,
                 subset_genes = ext_spatial_genes)

# cluster spatial genes
spat_cor_netw_DT = clusterSpatialCorGenes(spat_cor_netw_DT, name = &#39;spat_netw_clus&#39;, k = 20)

# visualize clusters
heatmSpatialCorGenes(visium_brain,
         spatCorObject = spat_cor_netw_DT,
         use_clus_name = &#39;spat_netw_clus&#39;,
         heatmap_legend_param = list(title = NULL),
         save_param = list(save_name=&quot;10_c_heatmap&quot;,
                   base_height = 6, base_width = 8, units = &#39;cm&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_c_heatmap.png"><img alt="10_c_heatmap.png" src="../../_images/10_c_heatmap.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>table(spat_cor_netw_DT$cor_clusters$spat_netw_clus)

coexpr_dt = data.table::data.table(genes = names(spat_cor_netw_DT$cor_clusters$spat_netw_clus),
           cluster = spat_cor_netw_DT$cor_clusters$spat_netw_clus)
data.table::setorder(coexpr_dt, cluster)
top30_coexpr_dt = coexpr_dt[, head(.SD, 30) , by = cluster]


# do HMRF with different betas on 500 spatial genes
my_spatial_genes &lt;- top30_coexpr_dt$genes

hmrf_folder = paste0(results_folder,&#39;/&#39;,&#39;11_HMRF/&#39;)
if(!file.exists(hmrf_folder)) dir.create(hmrf_folder, recursive = T)

HMRF_spatial_genes = doHMRF(gobject = visium_brain,
            expression_values = &#39;scaled&#39;,
            spatial_genes = my_spatial_genes, k = 20,
            spatial_network_name=&quot;spatial_network&quot;,
            betas = c(0, 10, 5),
            output_folder = paste0(hmrf_folder, &#39;/&#39;, &#39;Spatial_genes/SG_topgenes_k20_scaled&#39;))

visium_brain = addHMRF(gobject = visium_brain, HMRFoutput = HMRF_spatial_genes,
           k = 20, betas_to_add = c(0, 10, 20, 30, 40),
           hmrf_name = &#39;HMRF&#39;)

spatPlot(gobject = visium_brain, cell_color = &#39;HMRF_k20_b.40&#39;,
    point_size = 2, save_param=c(save_name=&quot;10_d_spatPlot2D_HMRF&quot;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/10_d_spatPlot2D_HMRF.png"><img alt="10_d_spatPlot2D_HMRF.png" src="../../_images/10_d_spatPlot2D_HMRF.png" style="width: 400px;" /></a>
</div>
</div>
<div class="section" id="export-and-create-giotto-viewer">
<h2>11. Export and Create Giotto Viewer<a class="headerlink" href="#export-and-create-giotto-viewer" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># check which annotations are available</span>
<span class="n">combineMetadata</span><span class="p">(</span><span class="n">visium_brain</span><span class="p">,</span> <span class="n">spat_enr_names</span> <span class="o">=</span> <span class="s1">&#39;PAGE&#39;</span><span class="p">)</span>

<span class="c1"># select annotations, reductions and expression values to view in Giotto Viewer</span>
<span class="n">viewer_folder</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">results_folder</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse_Visium_brain_viewer&#39;</span><span class="p">)</span>

<span class="n">exportGiottoViewer</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">visium_brain</span><span class="p">,</span>
       <span class="n">output_directory</span> <span class="o">=</span> <span class="n">viewer_folder</span><span class="p">,</span>
       <span class="n">spat_enr_names</span> <span class="o">=</span> <span class="s1">&#39;PAGE&#39;</span><span class="p">,</span>
       <span class="n">factor_annotations</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;in_tissue&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;HMRF_k20_b.40&#39;</span><span class="p">),</span>
       <span class="n">numeric_annotations</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;nr_genes&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;clus_25&#39;</span><span class="p">),</span>
       <span class="n">dim_reductions</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span><span class="p">),</span>
       <span class="n">dim_reduction_names</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span><span class="p">),</span>
       <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;scaled&#39;</span><span class="p">,</span>
       <span class="n">expression_rounding</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
       <span class="n">overwrite_dir</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>

        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;6_e_metaheatmap_scran&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="STARmap_mouse_cortex.html" class="btn btn-neutral float-left" title="STARmap Mouse Cortex" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="mouse_visium_kidney.html" class="btn btn-neutral float-right" title="Mouse Visium Kidney" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Ruben Dries, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>